### DISCUSSION TITLE
Support for esp32

### DISCUSSION THREAD

**Discussion by haikyuu** [slint-ui/slint#1077] [Q&A] [answered]
I am planning to build an ebook reader and so I ordered a LILYGO® E-ink Display
I am still looking for a convenient way to build the UI and this project seems like the most developer friendly one.

How hard is it to add support for this?

**Comment by ogoffart** [comment, marked as answer]
It all depends on the Rust support for it.
From my understanding after a quick search, rust should be supported without much problem on such devices.

The MCU port is still work in progress, but basically, to support a custom board, you have to implement a trait providing capabilities such as sending pixels to the screen and listening to the touchscreen.

This is the current trait (but it may change in the future):
https://github.com/slint-ui/slint/blob/eb9092ba15f212dc35f92b2bceae07b851cef6d2/internal/backends/mcu/lib.rs#L38-L48

So for example, this file implements the support for the RaspberryPi Pico  https://github.com/slint-ui/slint/blob/eb9092ba15f212dc35f92b2bceae07b851cef6d2/internal/backends/mcu/pico_st7789.rs
It contains all the code to configure the pins, the touch driver (which arguably should be in a different crate), a panic handler, and the implementation of the Devices trait.

(We will also expose that from other programming languages when we support that.)
